### PR TITLE
Scib artifact metrics

### PR DESCRIPTION
**Pull request by ergonyc**
adds two additional scripts to execute: 
- add_harmony.py :   make sure that we have PCA obsm available, and perform hamony integration.
- artifact_metrics : calculate scib metrics on the final dataset and export report

WDL changes #46 
- [ ] to create the output assessment report add calls to 
    - [ ] "add_harmony.py" and 
    - [ ] "artifact_metrics.py" 
- [ ] add "--output-validation-file"  arguments to 
    - [ ] "plot_qc_metrics.py",
    - [ ] "filter.py" and
    - [ ] "process.py"

### CODE REVIEW

File: docker/scvi/scripts/main/unused/annotate_clusters.py
Comment (kfang4) [inline docker/scvi/scripts/main/unused/annotate_clusters.py:1]: Remove `unused` folder

Comment (kfang4) [inline docker/scvi/scripts/main/unused/annotate_clusters.py:1]: Remove `unused` folder

File: docker/scvi/scripts/README.md [hunk 1]
```diff
@@ -1,38 +1,41 @@
+# scVI analysis Pipeline python scripts
 
 
-#### _PREPROCESS_
-- _pre-preprocessing_: `cellbender` ('cellbender.py') ~~or `scAR`('preprocess_scar.py').~~  (replacing `soupx` in preprocess.R)
-- _preprocessing_: `scrublet` ('preprocess.py') ~~or`SOLO` ('preprocess_solo.py')~~ 
+### _PREPROCESS_
+- _pre-preprocessing_: [`cellbender`](../../cellbender/scripts/cellbender.py)
```
Comment (kfang4) [inline docker/scvi/scripts/README.md:5]: Cellbender is called directly, not with script anymore

File: docker/scvi/scripts/main/annotate_cells.py [hunk 2]
```diff
@@ -8,101 +8,109 @@
...
 )
 parser.add_argument(
-    '--marker-genes',
-    dest='marker_genes',
+    "--marker-genes",
+    dest="marker_genes",
     type=str,
-    default='resources/celltype_marker_table.csv',
-    help='Path to marker_genes .csv file'
+    default="resources/celltype_marker_table.csv",
```
Comment (kfang4) [inline docker/scvi/scripts/main/annotate_cells.py:19]: Paths will be removed and the bucket uri will be used in WDL

File: docker/scvi/scripts/main/filter.py [hunk 3]
```diff
@@ -2,39 +2,54 @@
 import muon
 import scanpy
 import argparse
+import pandas as pd
+import pandas as pd
```
Comment (kfang4) [inline docker/scvi/scripts/main/filter.py:6]: Import duplicated

**Review by kfang4 — CHANGES_REQUESTED** [review]
@ergonyc reviewed. Which WDL tasks are going to use `add_harmony.py` and `artifact_metrics.py`?

File: docker/scvi/scripts/README.md (hunk 1 above)
Comment (ergonyc) [reply docker/scvi/scripts/README.md:5]: Right.  I will update the documentation noting that.

File: docker/scvi/scripts/main/annotate_cells.py (hunk 2 above)
Comment (ergonyc) [reply docker/scvi/scripts/main/annotate_cells.py:19]: default removed.

File: docker/scvi/scripts/main/artifact_metrics.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    scanpy.external.pp.harmony_integrate(adata, "sample")
+
+
+adata.obsm["Unintegrated"] = adata.obsm["X_pca"]
+
+biocons = BioConservation(isolated_labels=False)
+
+bm = Benchmarker(
+    adata,
+    batch_key="sample",
```
Comment (kfang4) [inline docker/scvi/scripts/main/artifact_metrics.py:88]: @ergonyc Is this supposed to hardcoded as sample? There's an unused argument `--batch-key` and `"batch_id"` has been used, see [process.py](https://github.com/ASAP-CRN/harmonized-wf-dev/blob/8c9c589523e5543af06222c662a7deecb02e5996/docker/scvi/scripts/main/process.py#L78-L83)

File: docker/scvi/scripts/main/artifact_metrics.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+bm.prepare(neighbor_computer=faiss_brute_force_nn)
+bm.benchmark()
+
+report_dir = Path.cwd() / args.scib_report_dir
+if not report_dir.exists():
+    report_dir.mkdir(parents=True, exist_ok=True)
+
+bm.plot_results_table(min_max_scale=False, save_dir=report_dir)
+df = bm.get_results(min_max_scale=False)
+df.to_csv((report_dir / "results.csv"), index=False)
```
Comment (kfang4) [inline docker/scvi/scripts/main/artifact_metrics.py:104]: @ergonyc does this file have to be named `results.csv` or can I change it to something more informative like `scib_report.csv`?

File: docker/scvi/scripts/main/artifact_metrics.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
+import argparse
+import scanpy
+import numpy as np
+from scib_metrics.benchmark import Benchmarker, BioConservation
+from scib_metrics.nearest_neighbors import NeighborsResults
+import faiss
+from pathlib import Path
+
+parser = argparse.ArgumentParser(description="Run scVI integration")
```
Comment (kfang4) [inline docker/scvi/scripts/main/artifact_metrics.py:9]: @ergonyc typo?

File: docker/scvi/scripts/main/add_harmony.py [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+# make sure we have PCA
+if "X_pca" not in adata.obsm:
+    scanpy.pp.pca(adata, n_comps=30)
+
+# add harmony
+if "X_pca_harmony" not in adata.obsm:
+    scanpy.external.pp.harmony_integrate(adata, args.batch_key)
+    # 9. write_h5ad
+
+adata.write_h5ad(filename=args.adata_output, compression="gzip")
```
Comment (kfang4) [inline docker/scvi/scripts/main/add_harmony.py:36]: @ergonyc Is this adata output used as input for the [plot_feats_and_groups.py](https://github.com/ASAP-CRN/harmonized-wf-dev/blob/scib_changes/docker/scvi/scripts/main/plot_feats_and_groups.py)?

File: docker/scvi/scripts/main/artifact_metrics.py (hunk 4 above)
Comment (ergonyc) [reply docker/scvi/scripts/main/artifact_metrics.py:88]: Yes we should remove the hard code. However for our current data the results are poor if sample is chosen (hence the hard code left over from testing).

File: docker/scvi/scripts/main/artifact_metrics.py (hunk 5 above)
Comment (ergonyc) [reply docker/scvi/scripts/main/artifact_metrics.py:104]: Yes. This should be changed.

File: docker/scvi/scripts/main/add_harmony.py (hunk 7 above)
Comment (ergonyc) [reply docker/scvi/scripts/main/add_harmony.py:36]: Yes, this is the "final" output. However, the current "features and groups" will not change from calling with the previous inputs. But logically it's the new "final artifact"

File: docker/scvi/scripts/main/artifact_metrics.py (hunk 4 above)
Comment (kfang4) [reply docker/scvi/scripts/main/artifact_metrics.py:88]: @ergonyc Thanks, so should I use `sample` or `batch_id` as `--batch-key`?

Comment (ergonyc) [reply docker/scvi/scripts/main/artifact_metrics.py:88]: Use "batch_id" as --batch_key

File: docker/scvi/scripts/main/artifact_metrics.py (hunk 6 above)
Comment (ergonyc) [reply docker/scvi/scripts/main/artifact_metrics.py:9]: Which line do you think is a typo?  .  The parser description ?   That is not correct.  I wlll change.

**Pull request merged**